overseer/src/lib.rs

```diff
@@ -0,0 +1,623 @@
+
+/// An asynchronous job that runs inside and being overseen by the `Overseer`.
+///
+/// In essence it's just a newtype wrapping a pinned `Future` dyn trait object.
+pub struct SubsystemJob(pub Pin<Box<dyn Future<Output = ()> + Send + 'static>>);
+
+/// A type of messages that are used inside the `Overseer`.
+///
+/// It is generic over some `M` that is intended to be a message type
+/// being used by the subsystems running on the `Overseer`. Most likely
+/// this type will be one large `enum` covering all possible messages in
+/// the system.
+/// It is also generic over `I` that is entended to be a type identifying
+/// different subsystems, again most likely this is a one large `enum`
+/// covering all possible subsystem kinds.
+enum OverseerMessage<M: Debug, I> {
+	/// This is a message generated by a `Subsystem`.
+	/// Wraps the messge itself and has an optional `to` of
+	/// someone who can receive this message.
+	///
+	/// If that `to` is present the message will be targetedly sent to the intended
+	/// receiver. The most obvious use case of this is communicating with children.
+	SubsystemMessage {
+		to: Option<I>,
+		msg: M,
+	},
+	/// A message that wraps something the `Subsystem` is desiring to
+	/// spawn on the overseer and a `oneshot::Sender` to signal the result
+	/// of the spawn.
+	SpawnChild {
+		s: (I, Box<dyn Subsystem<M, I> + Send>),
+		res: oneshot::Sender<Result<I, SubsystemError>>,
+	},
+}
+
+impl<M: Debug, I: Debug> Debug for OverseerMessage<M, I> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		match self {
+			OverseerMessage::SubsystemMessage { to, msg } => {
+				write!(f, "OverseerMessage::SubsystemMessage{{ to: {:?}, msg: {:?} }}", to, msg)
+			}
+			OverseerMessage::SpawnChild { .. } => write!(f, "OverseerMessage::Spawn(..)")
+		}
+	}
+}
+
+/// A running instance of some `Subsystem`.
+struct SubsystemInstance<M: Debug, I> {
+	/// We talk to the `Overseer` over this channel.
+	rx: mpsc::Receiver<OverseerMessage<M, I>>,
+	/// The `Overseer` talks to use over this channel.
+	tx: mpsc::Sender<M>,
+}
+
+/// An `id` that is given to any `SubsystemInstance` for identification.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct SubsystemId(usize);
+
+/// A context type that is given to the `Subsystem` upon spawning
+/// that can be used by `Subsystem` to communicate with the outside world.
+pub struct SubsystemContext<M: Debug, I>{
+	rx: mpsc::Receiver<M>,
+	tx: mpsc::Sender<OverseerMessage<M, I>>,
+}
+
+impl<M: Debug, I> SubsystemContext<M, I> {
+	/// Try to asyncronously receive a message.
+	///
+	/// This has to be used with caution, if you loop over this without
+	/// using `pending!()` macro you will end up with a busy loop!
+	pub async fn try_recv(&mut self) -> Result<Option<M>, ()> {
+		match poll!(self.rx.next()) {
+			Poll::Ready(Some(msg)) => Ok(Some(msg)),
+			Poll::Ready(None) => Err(()),
+			Poll::Pending => Ok(None),
+		}
+	}
```

Comment:
Why make this an `async` function without any `await`s? Why not return `Poll<Option<M>>` here? Why deviate from the ecosystem standard that something returns `Poll::Pending` when not yet ready?

```suggestion
	pub fn try_recv(&mut self, cx: Context<_>) -> Poll<Option<M>> {
		self.rx.poll_next_unpin(cx)
	}
```

Reply:
We want to call it from the `async` context:
```rust
async {
    ctx.try_recv().await;
}
```
So we don't actually want the user to deal with `Context` since that's not something that's being passed around, rather we have to somehow get it from the environment; so that is hidden by the `poll!` macro and we in turn hide it's usage from the user by this method. But the method needs to be `async` for `poll!` to work.

Also it would make no sense to use `Poll` in the `async` context.

Does this explanation make any sense at all? But the return type probably could be made a bit better

Reply:
As a user I would be confused to `await` a function that explicitly returns directly (`try` prefix). Instead I would find `poll!(ctx.try_recv())` easier to understand as it implies that the call logically returns directly.

Anyways, this is not a big deal, if no one else feels the same this should stay as it is.